sklearn/decomposition/tests/test_nmf.py

```diff
@@ -258,3 +287,228 @@ def test_safe_compute_error():
+            l1_ratio=l1_ratio, regularization='both')
+
+        assert_array_almost_equal(W1, W3, decimal=4)
+        assert_array_almost_equal(H1, H3, decimal=4)
+
+
+@ignore_warnings(category=DeprecationWarning)
+def test_nmf_regularization():
+    # Test the effect of L1 and L2 regularizations
+    n_samples = 6
+    n_features = 5
+    n_components = 3
+    rng = np.random.mtrand.RandomState(42)
+    X = np.abs(rng.randn(n_samples, n_features))
+
+    # L1 regularization should increase the number of zeros
+    l1_ratio = 1.
+    for solver in ['pg', 'cd', 'mu']:
+        regul = nmf.NMF(n_components=n_components, solver=solver,
+                        alpha=0.5, l1_ratio=l1_ratio)
+        model = nmf.NMF(n_components=n_components, solver=solver,
+                        alpha=0., l1_ratio=l1_ratio)
+
+        W_regul = regul.fit_transform(X)
+        W_model = model.fit_transform(X)
+
+        H_regul = regul.components_
+        H_model = model.components_
+
+        W_regul_n_zeros = W_regul[W_regul == 0].size
+        W_model_n_zeros = W_model[W_model == 0].size
+        H_regul_n_zeros = H_regul[H_regul == 0].size
+        H_model_n_zeros = H_model[H_model == 0].size
+
+        assert_greater(W_regul_n_zeros, W_model_n_zeros)
+        assert_greater(H_regul_n_zeros, H_model_n_zeros)
+
+    # L2 regularization should decrease the mean of the coefficients
+    l1_ratio = 0.
+    for solver in ['pg', 'cd', 'mu']:
+        regul = nmf.NMF(n_components=n_components, solver=solver,
+                        alpha=0.5, l1_ratio=l1_ratio)
+        model = nmf.NMF(n_components=n_components, solver=solver,
+                        alpha=0., l1_ratio=l1_ratio)
+
+        W_regul = regul.fit_transform(X)
+        W_model = model.fit_transform(X)
+
+        H_regul = regul.components_
+        H_model = model.components_
+
+        assert_greater(W_model.mean(), W_regul.mean())
+        assert_greater(H_model.mean(), H_regul.mean())
+
+
+# remove DeprecationWarning and UserWarning after removing 'pg' solver
+@ignore_warnings(category=(DeprecationWarning,
+                           ConvergenceWarning,
+                           UserWarning))
+def test_nmf_decreasing():
+    # test that the objective function is decreasing at each iteration
+    n_samples = 20
+    n_features = 15
+    n_components = 10
+    alpha = 0.1
+    l1_ratio = 0.5
+    tol = 0.
+
+    # initialization
+    rng = np.random.mtrand.RandomState(42)
+    X = rng.randn(n_samples, n_features)
+    np.abs(X, X)
+    W0, H0 = nmf._initialize_nmf(X, n_components, init='random',
+                                 random_state=42)
+
+    for beta_loss in (-1.2, 0, 0.2, 1., 2., 2.5):
+        for solver in ('pg', 'cd', 'mu'):
+            if solver != 'mu' and beta_loss != 2:
+                # not implemented
+                continue
+            W, H = W0.copy(), H0.copy()
+            previous_loss = None
+            for _ in range(1, 5):
```

Comment:
I added `nls_max_iter=20` since the bottleneck is the 'pg' solver.
With this option, we can go up to 30 iterations in 300ms.